docs/source/working/data.rst

```diff
@@ -0,0 +1,148 @@
+
+.. code:: python
+
+    from aiida.orm import Data
+
+    class NewData(Data)
+        """A new data type that wraps a single value."""
+
+        attribute_name = 'value'
+
+        def __init__(self, **kwargs)
+            value = kwargs.pop(self.attribute_name)
+            super(NewData, self).__init__(**kwargs)
+            self.set_attribute(self.attribute_name, value)
+
+By adding the value to the node's attributes, after the node itself has been stored, it will be permanently recorded in the database.
+Creating an instance of the ``NewData`` data type and storing it is now very easy:
+
+.. code:: python
+
+    node = Data(value=5)
+    node.store()
+
+By storing the ``value`` in the attributes of the node instance, we ensure that that ``value`` can be retrieved even when the node is reloaded at a later point in time.
+
+Besides making sure that the content of a data node is stored in the database or file repository, the data type class can also provide useful methods for users to retrieve that data.
+For example, with the current state of the ``NewData`` class in the last example, if a user were to load a stored ``NewData`` node and wanted to retrieve the ``value`` that it contains, they would have to do something like:
+
+.. code:: python
+
+    node = load_node(<IDENTIFIER>)
+    node.get_attribute(NewData.attribute_name)
+
+In other words, the user of the ``NewData`` class needs to know that the ``value`` is stored as an attribute with a name given by the class-attribute ``attribute_name``.
+This is not easy to remember and therefore not very user-friendly.
+Since the ``NewData`` type is a class, we can give it useful methods.
+Let's introduce one that will return the value that was stored for it:
+
+.. code:: python
+
+    from aiida.orm import Data
+
+    class NewData(Data)
+        """A new data type that wraps a single value."""
+
+        attribute_name = 'value'
+
+        @property
+        def value(self):
+            """Return the value stored for this instance."""
+            return self.get_attribute(self.attribute_name)
+
+The addition of the instance property ``value`` makes retrieving the value of a ``NewData`` node a lot easier:
+
+.. code:: python
+
+    node = load_node(<IDENTIFIER)
+    value = node.value
+
+The new data type we have just implemented is of course trivial and not very useful, but the concept is flexible and can be easily applied to more complex use-cases.
+The following section will provide useful guidelines on how to optimally design new data types.
+
+
+.. _working_data_design_guidelines:
+
+Design guidelines
+=================
+
+Database or repository?
+-----------------------
+
+When deciding where to store a property of a data type, one has to choose between the database and the file repository.
+The database will make the property queryable and relatively quickly accessible.
+The downside is that storing too much information in the database can make it sluggish.
```

Comment:
Either I don't understand this phrase or it is incorrect. A database surely does not have to fit in memory in its entirety? It is just the bigger it becomes the more data will have to be considered when looking for something specific (in lay-mans terms).

Reply:
I don't know when we wrote the initial sentence, anyway too much info in the DB can make it slower not because information do not fit in memory (or, better, only indirectly because of this) because of the way data is loaded into memory while querying (I think it's loaded in "pages", and if a row contains a lot of data (e.g. in extras or attributes) a page will contain very few rows.

Another reason is that if the content is indexed, more content means slower insertion (because of indexing) and a bit slower use of the indexes.

Another issue that maybe instead one can also point out is that the DB anyway adds a layer of complexity, so for "raw" data (files) it's not worth putting in the DB.

An even additional reason is that if the DB is small, it's easier to backup (for files there are incremental ways like rsync to backup it, for DB we typically do periodically a full dump).

I don't know how much of this info we want to mention here (maybe can be put in a note?)

Reply:
Feel free to dismiss my change and update according to your liking